FWCore/Framework/interface/ProxyArgumentFactoryTemplate.h

```diff
@@ -21,43 +21,45 @@
 // system include files
 #include <memory>
 #include <string>
+#include <vector>
 
 // user include files
 #include "FWCore/Framework/interface/ProxyFactoryBase.h"
 #include "FWCore/Framework/interface/DataKey.h"
 
-// forward declarations
 namespace edm {
   namespace eventsetup {
 
-    template <class T, class ArgT>
+    class DataProxy;
+
+    template <class ProxyType, class CallbackType>
     class ProxyArgumentFactoryTemplate : public ProxyFactoryBase {
     public:
-      typedef typename T::record_type record_type;
-
-      ProxyArgumentFactoryTemplate(ArgT iArg) : arg_(iArg) {}
-      //virtual ~ProxyArgumentFactoryTemplate()
-
-      // ---------- const member functions ---------------------
-      std::unique_ptr<DataProxy> makeProxy() const override { return std::make_unique<T>(arg_); }
+      using RecordType = typename ProxyType::RecordType;
+
+      ProxyArgumentFactoryTemplate(std::shared_ptr<std::vector<std::shared_ptr<CallbackType>>> callbacks)
+          : callbacks_(std::move(callbacks)) {}
+
+      ProxyArgumentFactoryTemplate(const ProxyArgumentFactoryTemplate&) = delete;
+      const ProxyArgumentFactoryTemplate& operator=(const ProxyArgumentFactoryTemplate&) = delete;
+
+      std::unique_ptr<DataProxy> makeProxy(unsigned int iovIndex) override {
+        while (iovIndex >= callbacks_->size()) {
+          callbacks_->push_back(std::make_shared<CallbackType>(callbacks_->at(0)->get(),
+                                                               callbacks_->at(0)->method(),
+                                                               callbacks_->at(0)->transitionID(),
+                                                               callbacks_->at(0)->decorator()));
+        }
+        return std::make_unique<ProxyType>(callbacks_->at(iovIndex));
+      }
 
       DataKey makeKey(const std::string& iName) const override {
-        return DataKey(DataKey::makeTypeTag<typename T::value_type>(), iName.c_str());
+        return DataKey(DataKey::makeTypeTag<typename ProxyType::ValueType>(), iName.c_str());
       }
 
-      // ---------- static member functions --------------------
-
-      // ---------- member functions ---------------------------
-
     private:
-      ProxyArgumentFactoryTemplate(const ProxyArgumentFactoryTemplate&) = delete;  // stop default
-
-      const ProxyArgumentFactoryTemplate& operator=(const ProxyArgumentFactoryTemplate&) = delete;  // stop default
-
-      // ---------- member data --------------------------------
-      mutable ArgT arg_;
+      std::shared_ptr<std::vector<std::shared_ptr<CallbackType>>> callbacks_;
```

Comment:
Is there actually any reason for `ProxyArgumentFactoryTemplate` to hold onto all the callbacks? Isn't it just a factory so all it needs is a way to create new callbacks?

Reply:
I think I've figured out my own question. Since one Callback can be used for multiple DataProxies (if a produces method returns multiple items) the factory needs to hold onto the Callback between calls.

Reply:
Callback has IOV dependent state. So there needs to be one per IOV. In ESProducer.h in the constructor, we do not know how many IOVs there will be yet, but we know we will need at least one. So one is constructed there. Part of this is convenience because the code to build the first Callback was already in ESProducer.h. It is fairly complicated template code and I just made minimal changes to it.  I am not sure how difficult it would be to move this into the factory. This would take some thought. After the first Callback is made it is fairly easy to copy a shared_ptr into the factory and copy the Callback to make the number we need when we know how many we need. I'm open to reorganizing this if you would like. I am still looking at it to think about how it could be done better.

Reply:
@Dr15Jones Here is one alternative, in Callback.h change the data
members of Callback as follows:

std::array<void*, produce::size<TReturn>::value> proxyData_;
to
std::vector<std::array<void*, produce::size<TReturn>::value>> proxyData_;

and also make wasCalledForThisRecord_ a std::vector<bool>

Then I can pass around the iovIndex to get the proper element
in the new vectors. This adds one more level of indirection, but
does simplify the code that initializes things. Only one Callback
object would be needed instead of a vector of them.
Does that sound any better?

Reply:
As we discussed offline, the answer to the original question is yes because the same Callback is used in multiple factories, but we need a different one for each IOV. I pushed a commit to store these in a pair containing the iovIndex and a shared_ptr<Callback>.